dsgrid/config/time_dimension_base_config.py

```diff
@@ -110,3 +123,82 @@ def list_expected_dataset_timestamps(self):
 
         """
+
+    def _convert_time_to_project_time_interval(
+        self, df=None, project_time_dim=None, wrap_time=True
+    ):
+        """Shift time to match project time based on TimeIntervalType
+        If time range spills over into another year after time interval alignment,
+        the time range will be wrapped around so it's bounded within the year
+        """
+        if project_time_dim is None:
+            return df
+
+        dtime_interval = self.get_time_interval_type()
+        ptime_interval = project_time_dim.get_time_interval_type()
+        time_col = project_time_dim.get_timestamp_load_data_columns()
+
+        assert len(time_col) == 1, time_col
+        time_col = time_col[0]
+
+        if dtime_interval != ptime_interval:
+            match (dtime_interval, ptime_interval):
+                case (TimeIntervalType.PERIOD_BEGINNING, TimeIntervalType.PERIOD_ENDING):
+                    df = df.withColumn(
+                        time_col,
+                        F.col(time_col)
+                        + F.expr(f"INTERVAL {self.get_frequency().seconds} SECONDS"),
+                    )
+                case (TimeIntervalType.PERIOD_ENDING, TimeIntervalType.PERIOD_BEGINNING):
+                    df = df.withColumn(
+                        time_col,
+                        F.col(time_col)
+                        - F.expr(f"INTERVAL {self.get_frequency().seconds} SECONDS"),
+                    )
+
+        if wrap_time:
+            df = self._apply_time_wrap(df, project_time_dim)
+
+        return df
+
+    @staticmethod
+    def _apply_time_wrap(df, project_time_dim):
+        """If dataset_time does not match project_time, apply time-wrapping"""
+
+        time_col = project_time_dim.get_timestamp_load_data_columns()
+        assert len(time_col) == 1, time_col
+        time_col = time_col[0]
+
+        project_time = set(
+            project_time_dim.build_time_dataframe().agg(F.collect_set(time_col)).collect()[0][0]
+        )
+        dataset_time = set(df.agg(F.collect_set(time_col)).collect()[0][0])
+        diff = dataset_time.difference(project_time)
+
+        if not diff:
+            return df
+
+        # extract time_delta based on if diff is to the left or right of project_time
+        time_delta = (
+            max(project_time) - min(project_time) + project_time_dim.get_frequency()
+        ).total_seconds()
+        if min(diff) > max(project_time):
+            time_delta *= -1
+
+        # time-wrap by "changing" the year with time_delta
+        df = (
+            df.filter(F.col(time_col).isin(diff))
+            .withColumn(
+                time_col,
+                F.from_unixtime(F.unix_timestamp(time_col) + time_delta).cast("timestamp"),
+            )
+            .union(df.filter(~F.col(time_col).isin(diff)))
+        )
+
+        dataset_time = set(df.agg(F.collect_set(time_col)).collect()[0][0])
```

Comment:
`set(df.select(time_col).distinct().collect())` is more straightforward and is likely a better solution.